### PR TITLE
1536 utxo regression test

### PR DESCRIPTION
**Pull request by kayvank**
# DO NOT MERGE
---------------------------
Regression test for Uxo indexers.

## Note:
We expect this test to fail in this branch:
```
cabal test marconi-chain-index --test-option=--pattern="marconi-utxo storage-query"
```
We expect the same test to pass in [PR-1021](https://github.com/input-output-hk/plutus-apps/pull/1021)

## TODO
PR should be closed once PR-1021 is merged.

[PLT-1536][PLT-1540]

[PLT-1536]: https://input-output.atlassian.net/browse/PLT-1536?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 1]
```diff
@@ -120,24 +126,43 @@ utxoStorageTest = property $ do
...
+{-|
+  Insert Utxo events in storage, and retrieve the events
+  The property we're checking here is:
+    - retreived at least one unspent utxo
+    - only retreive unspent utxo's
+    - test `spent` filtering at the boundry of in-memory & disk storage
+-}
+propSaveAndRetrieveUtxoEvents :: Property
+propSaveAndRetrieveUtxoEvents = property $ do
+  -- events <- forAll genUtxoEvents'' -- TODO
```
Comment (kayvank) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:139]: this generator pushes all the spent to the last element of the [events], thus making sure we filter out the in-memory spent. Unfortunately, it meant we weren't testing the SQL LEFT join, therefor I opted out in using it.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo/Types.hs [hunk 2]
```diff
@@ -0,0 +1,260 @@
+{-# LANGUAGE AllowAmbiguousTypes   #-}
```
Comment (koslambrou) [inline marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo/Types.hs:1]: I don't think doing a Types module is good idea like I previously mentioned.

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 3]
```diff
@@ -120,24 +126,43 @@ utxoStorageTest = property $ do
...
---                    (filter (\e -> Utxo.ueChainPoint e /= C.ChainPointAtGenesis) events) )
+---------------- Regression test -------------------
+{-|
+  Insert Utxo events in storage, and retrieve the events
+  The property we're checking here is:
+    - retreived at least one unspent utxo
+    - only retreive unspent utxo's
+    - test `spent` filtering at the boundry of in-memory & disk storage
+-}
+propSaveAndRetrieveUtxoEvents :: Property
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:137]: I'm having a hard time understand what this tests exactly. Or better yet, isn't there another propery test for querying which could have been adapted to detect this regression?

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs (hunk 1 above)
Comment (koslambrou) [reply marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:139]: Why filter out the in-memory spent?

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 4]
```diff
@@ -120,24 +126,43 @@ utxoStorageTest = property $ do
...
+propSaveAndRetrieveUtxoEvents = property $ do
+  -- events <- forAll genUtxoEvents'' -- TODO
+  events <- forAll genShelleyEraUtxoEvents
+  let numOfEvents = length events
+  depth <- forAll $ Gen.int (Range.constantFrom (numOfEvents - 1) 1 (numOfEvents + 1))
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:142]: Maybe a comment here explaining why you choose that range of depth?

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 5]
```diff
@@ -120,24 +126,43 @@ utxoStorageTest = property $ do
...
+        = rowsFromStorage & each %~ (\r ->
+                                       C.TxIn (r ^. Utxo.urUtxo . Utxo.txId)(r ^. Utxo.urUtxo . Utxo.txIx))
+      fromEventsTxIns :: [C.TxIn]
+        = concatMap (\(Utxo.UtxoEvent _ ins _ ) -> Set.toList ins ) events
+
+  Hedgehog.classify "Query both in-memory and storage " $ depth < numOfEvents
+  Hedgehog.classify "Query in-memory only" $ depth > numOfEvents
+
+  -- A property of the generator is that there is at least one unspent transaction
+  Hedgehog.assert (not . null $ rowsFromStorage)
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:160]: How are you sure of that, and how is that relevant for the property test?

**Review by koslambrou — COMMENTED** [review]
Ok, I'm having a hard time understanding the property test and it's preconditions :(

Let me put in some time tomorrow to figure out a simpler property test for that.

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs (hunk 1 above)
Comment (kayvank) [reply marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs]: I am closing this per our discussion @koslambrou . Also added additional detail to:  [Spec.Marconi.ChainIndex.Indexers.Utxo.UtoxIndex.hs](https://github.com/kayvank/plutus-apps/blob/kayvan/1536-utxo-regression-test/marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs#L90)

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs (hunk 3 above)
Comment (kayvank) [reply marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs]: added detail to the test, please see: [Spec.Marconi.ChainIndex.Indexers.Utxo.UtxoIndex ](https://github.com/kayvank/plutus-apps/blob/kayvan/1536-utxo-regression-test/marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs#L90)
